### PR TITLE
Assert that an element is visible to the user.

### PR DESCRIPTION
**Pull request by aserv92**
bug #169

Method:
The "I should see assertion"

Expected Behavior:
Asserts that the user can see text on the screen.

Actual Behavior:
Asserts that text is anywhere in the DOM, but not necessarily visible to the user.

Cause:
The "I should see assertion" checks the DOM for text

Solution:
Check the position and visibility of all parent elements of a specified element that contains the text you are looking for.

Example:

"Test Text" should be fully displayed

<span data-qa-id="Test Text">Test Text</span>

### CODE REVIEW

**Comment by healthlabs-bugman** [comment]
Paging @juanmrad to review the front-end work in this pull request!
Paging @ramsey-medology to review the back-end work in this pull request!
Paging @Chekote to review the testing work in this pull request!

**Comment by aserv92** [comment]
!reassign testing @jose-ochoa

**Comment by healthlabs-bugman** [comment]
Paging @jose-ochoa to review the testing work in this pull request!

**Comment by aserv92** [comment]
!reassign front-end @juanmrad

**Comment by healthlabs-bugman** [comment]
Paging @juanmrad to review the front-end work in this pull request!

**Comment by aserv92** [comment]
!reassign back-end @fortoulrodolfo

**Comment by healthlabs-bugman** [comment]
Paging @fortoulrodolfo to review the back-end work in this pull request!

**Comment by aserv92** [comment]
Manually reassigned reviewers to match PR#186 because both versions needed to be updated with the same methods.

File: src/Medology/Behat/Mink/FlexibleContext.php [hunk 1]
```diff
@@ -1240,4 +1241,167 @@ public function waitForPageLoad()
...
+     * @throws UnsupportedDriverActionException
+     * @return Rectangle                        representing the viewport
+     */
+    public function getElementViewportRectangle(NodeElement $element)
+    {
+        $driver = $this->getSession()->getDriver();
+
+        if (!$driver instanceof Selenium2Driver) {
+            throw new UnsupportedDriverActionException('%s does not support assertNodeIsFullyVisibleInViewPort', $driver);
+        }
```
Comment (Chekote) [inline src/Medology/Behat/Mink/FlexibleContext.php]: FYI: It wasn't available at the time you wrote this, but there's now a method called assertSelenium2Driver that you can use to simplify this.

File: src/Medology/Behat/Mink/FlexibleContext.php [hunk 2]
```diff
@@ -1240,4 +1241,167 @@ public function waitForPageLoad()
...
+        $NodeElements = [];
+        while ($NodeElement->getParent() instanceof NodeElement) {
+            $NodeElements[] = ($NodeElement = $NodeElement->getParent());
+            if (strtolower($NodeElement->getTagName()) === strtolower($stopAt)) {
+                break;
+            }
+        }
+        if ($reverseOrder) {
+            $NodeElements = array_reverse($NodeElements);
+        }
```
Comment (Chekote) [inline src/Medology/Behat/Mink/FlexibleContext.php]: In the interests of SRP I would not offer this feature within this function. If the caller wants the list reversed, calling array_reverse themselves is trivial.

File: src/Medology/Behat/Mink/FlexibleContext.php [hunk 3]
```diff
@@ -1240,4 +1241,167 @@ public function waitForPageLoad()
...
+
+    /**
+     * Get list of of all NodeElement parents.
+     *
+     * @param  NodeElement $NodeElement
+     * @param  string      $stopAt       html tag to stop at
+     * @param  bool        $reverseOrder list parents in reverse order (root element will be at index 0)
+     * @return array       of nodeElements
+     */
+    private function getListOfAllNodeElementParents(NodeElement $NodeElement, $stopAt, $reverseOrder)
```
Comment (Chekote) [inline src/Medology/Behat/Mink/FlexibleContext.php]: Please use lower camel case for all variables `$nodeElement` vs `$NodeElement`. Upper camel case is reserved for classes, traits, interfaces, etc.

File: web/ScrollableDivs.html [hunk 4]
```diff
@@ -0,0 +1,307 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Scrollable Divs</title>
+    <script>
+      /**
+       *
```
Comment (Chekote) [inline web/ScrollableDivs.html]: Please remove this blank line.

File: web/ScrollableDivs.html [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+       * Centers all scrolls
+       *
+       * @param ClassName
+       * @param x x offset from center
+       * @param y y offset from center
```
Comment (Chekote) [inline web/ScrollableDivs.html]: Please add types for these parameters. See [jsdoc](http://usejsdoc.org/tags-param.html).

**Pull request merged**